internals/overlord/servstate/manager.go

```diff
@@ -88,6 +88,9 @@ func (m *ServiceManager) Stop() {
 	if err != nil {
 		logger.Noticef("Cannot stop child process reaper: %v", err)
 	}
+	for name := range m.services {
```

Comment:
I'm not quite sure what you mean by "servicelock" - do you mean `m.servicesLock`?

Since `m.services` can be accessed concurrently, we should lock `m.servicesLock` whenever we access it. This locking is done inside the `removeService` method.

Even if it's not possible for anything to concurrently access `m.services` here, it still doesn't hurt to lock it. It's safe, adds negligible overhead, and future-proofs against race conditions.

Does that address your concern?

Reply:
reading m.services needs the servicesLock, since removeService needs the lock too there are probably two options:

Copy the names:
```
// Stop implements overlord.StateStopper and stops background functions.
func (m *ServiceManager) Stop() {
	err := reaper.Stop()
	if err != nil {
		logger.Noticef("Cannot stop child process reaper: %v", err)
	}
	m.servicesLock.Lock()
	services := make([]string, len(m.services))
	for name := range m.services {
		services = append(services, name)
	}
	m.servicesLock.Unlock()
	for _, name := range services {
		m.removeService(name)
	}
}
```

Add a removeAllServices method:
```
// Stop implements overlord.StateStopper and stops background functions.
func (m *ServiceManager) Stop() {
	err := reaper.Stop()
	if err != nil {
		logger.Noticef("Cannot stop child process reaper: %v", err)
	}
	m.removeAllServices()
}

func (m *ServiceManager) removeAllServices() {
	m.servicesLock.Lock()
	defer m.servicesLock.Unlock()

	for name, svc := range m.services {
		if svc.logs != nil {
			err := svc.logs.Close()
			if err != nil {
				logger.Noticef("Error closing service %q ring buffer: %v", name, err)
			}
		}
	}

	m.services = map[string]*serviceData{}
}
```

Reply:
ah I see the issue now. Sorry @flotter I misunderstood your comment